Patches to run cloudbase-init plugins on Solaris.  Possibly upstreamable

--- cloudbase-init-1.1.4/cloudbaseinit/osutils/posix.py
+++ cloudbase-init-1.1.4/cloudbaseinit/osutils/posix.py
@@ -12,12 +12,106 @@
 #    License for the specific language governing permissions and limitations
 #    under the License.
 
+import crypt
+import ctypes as C
 import os
+import pwd
 
+from subprocess import CalledProcessError, Popen, PIPE, check_call
 from cloudbaseinit.osutils import base
 
 
+def encrypt_password(plaintext):
+
+    class Passwd(C.c_void_p):
+        # Since we are passing NULL as the second argument,
+        # all we need is a void pointer
+        pass
+
+    _libc = C.CDLL("libc.so.1")
+
+    crypt_gensalt = _libc.crypt_gensalt
+    crypt_gensalt.argtypes = (C.c_char_p, C.POINTER(Passwd))
+    crypt_gensalt.restype = C.c_char_p
+
+    salt = crypt_gensalt(None, None)
+    return crypt.crypt(plaintext, salt.decode())
+
+
 class PosixUtils(base.BaseOSUtils):
 
+    def __init__(self):
+        self._config_value = {}
+
     def reboot(self):
         os.system('reboot')
+
+    def get_config_value(self, name, section=None):
+        """Get Plugin Status for a (plugin name, section) pair.
+        For example: PLUGIN_EXECUTION_DONE"""
+
+        if not section:
+            config_key = name
+        else:
+            config_key = name + ":" + section
+
+        return self._config_value.get(config_key)
+
+    def set_config_value(self, name, value, section=None):
+        """Set Plugin Status for a (plugin name, section) pair.
+        For example: PLUGIN_EXECUTION_DONE"""
+
+        if not section:
+            config_key = name
+        else:
+            config_key = name + ":" + section
+
+        self._config_value[config_key] = value
+
+    def set_host_name(self, new_host_name):
+        """Set current instance's hostname"""
+
+        try:
+            check_call(["/usr/bin/hostname", new_host_name])
+            return False # Reboot not required
+        except CalledProcessError:
+            raise RuntimeError("Unable to set host name")
+
+    def change_password_next_logon(self, username):
+        """Force user to change password during next login."""
+
+        check_call(["/usr/bin/passwd", "-f", username])
+
+    def set_user_password(self, username, password, password_expires=False):
+        """Change password for a specific user"""
+
+        hashed_password = encrypt_password(password)
+        check_call(["/usr/bin/passwd", "-p", hashed_password, username])
+
+    def add_user_to_local_group(self, username, groupname):
+        """Add user to a specific local group"""
+
+        check_call(["/usr/sbin/usermod", "-G", groupname, username])
+
+    def user_exists(self, username):
+        """Return True if user exists. """
+
+        try:
+            pwd.getpwnam(username)
+            return True
+        except KeyError:
+            return False
+
+    def get_user_home(self, username):
+        """Get user's home directory"""
+        
+        try:
+            d = pwd.getpwnam(username).pw_dir
+            return d
+        except KeyError:
+            return None
+
+    def get_maximum_password_length(self):
+        """Solaris maximum password length is 255"""
+
+        return 255
